### PR TITLE
MMP-314 The MuleVersion element should validate the minimum runtime v…

### PR DESCRIPTION
**Pull request by WilliamPinho**
…ersion for the application being deployed in all deployment targets

### CODE REVIEW

File: mule-deployer/src/main/java/org/mule/tools/client/agent/AgentClient.java [hunk 1]
```diff
@@ -57,4 +58,22 @@ public void undeployApplication(String appName) {
     }
   }
 
+  public String getMuleRuntimeVersion() {
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/client/agent/AgentClient.java]: wouldn't it make more sense to just have a getAgentInfo and return the AgentInfo POJO? 

Currently we have no reason for this to be a problem, and granted we've done this in other clients BUT in this specific scenario it kind of sound better to just return AgentInfo and have the user to get the data from it wouldn't you say ?

Comment (damiansima) [reply mule-deployer/src/main/java/org/mule/tools/client/agent/AgentClient.java]: also if you agree please create getters and setters for the AgentInfo POJO (we'll do the same with the other ones in due time)

File: mule-deployer/src/main/java/org/mule/tools/validation/DeploymentEnvironmentVersion.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  /**
+   * Checks if two versions are the same. The versions are supposed to follow the semantic versioning specification.
+   * 
+   * @param version1
+   * @param version2
+   * @return true if version1 is equal to version2; false otherwise.
+   */
+  private boolean isSameVersion(String version1, String version2) {
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/validation/DeploymentEnvironmentVersion.java]: we should have this logic in versionsUtils and proxy it from here

File: mule-deployer/src/main/java/org/mule/tools/validation/DeploymentValidatorFactory.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import org.mule.tools.model.standalone.StandaloneDeployment;
+import org.mule.tools.validation.agent.AgentDeploymentValidator;
+import org.mule.tools.validation.arm.ArmDeploymentValidator;
+import org.mule.tools.validation.cloudhub.CloudHubDeploymentValidator;
+import org.mule.tools.validation.standalone.StandaloneDeploymentValidator;
+
+/**
+ * A factory of {@link AbstractDeploymentValidator}.
+ */
+public class DeploymentValidatorFactory {
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/validation/DeploymentValidatorFactory.java]: Please add a test for this factory

File: mule-deployer/src/main/java/org/mule/tools/validation/agent/AgentDeploymentValidator.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import org.mule.tools.client.standalone.exception.DeploymentException;
+import org.mule.tools.model.Deployment;
+import org.mule.tools.model.agent.AgentDeployment;
+import org.mule.tools.validation.AbstractDeploymentValidator;
+import org.mule.tools.validation.DeploymentEnvironmentVersion;
+
+/**
+ * Validates if the mule runtime version is valid in an Agent deployment scenario.
+ */
+public class AgentDeploymentValidator extends AbstractDeploymentValidator {
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/validation/agent/AgentDeploymentValidator.java]: Please add a test for this class

File: mule-deployer/src/main/java/org/mule/tools/validation/arm/ArmDeploymentValidator.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import org.mule.tools.client.standalone.exception.DeploymentException;
+import org.mule.tools.model.Deployment;
+import org.mule.tools.model.anypoint.ArmDeployment;
+import org.mule.tools.validation.AbstractDeploymentValidator;
+import org.mule.tools.validation.DeploymentEnvironmentVersion;
+
+/**
+ * Validates if the mule runtime version is valid in an ARM deployment scenario.
+ */
+public class ArmDeploymentValidator extends AbstractDeploymentValidator {
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/validation/arm/ArmDeploymentValidator.java]: please add a test for this class

File: mule-deployer/src/main/java/org/mule/tools/validation/cloudhub/CloudHubDeploymentValidator.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import org.mule.tools.validation.AbstractDeploymentValidator;
+import org.mule.tools.validation.DeploymentEnvironmentVersion;
+
+import java.io.IOException;
+import java.util.Set;
+
+/**
+ * Validates if the mule runtime version is valid in a CloudHub deployment scenario.
+ */
+public class CloudHubDeploymentValidator extends AbstractDeploymentValidator {
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/validation/cloudhub/CloudHubDeploymentValidator.java]: please add a test for this class

File: mule-deployer/src/main/java/org/mule/tools/validation/standalone/MuleCoreJarVersionFinder.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+   */
+  private String muleCoreVersion;
+
+  @Override
+  public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+    return shouldContinueTraversing();
+  }
+
+  @Override
+  public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/validation/standalone/MuleCoreJarVersionFinder.java]: please add a test for this class

File: mule-deployer/src/main/java/org/mule/tools/validation/standalone/StandaloneDeploymentValidator.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+import java.io.File;
+import java.io.IOException;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.nio.file.Files.walkFileTree;
+
+/**
+ * Validates if the mule runtime version is valid in a Standalone deployment scenario.
+ */
+public class StandaloneDeploymentValidator extends AbstractDeploymentValidator {
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/validation/standalone/StandaloneDeploymentValidator.java]: please add a test for this class

File: mule-deployer/src/main/java/org/mule/tools/validation/DeploymentEnvironmentVersion.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.Collection;
+import java.util.List;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+/**
+ * Represents the supported mule runtime versions in a deployment environment.
+ */
+public class DeploymentEnvironmentVersion {
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/validation/DeploymentEnvironmentVersion.java]: could we name this class EnvironmentSupportedVersions 
The fact that is for the deployment doesn't seem to be that relevant
Name it Version(S) highlight the fact that there may be multiple of them

File: mule-deployer/src/main/java/org/mule/tools/client/agent/AgentClient.java (hunk 1 above)
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/client/agent/AgentClient.java]: Done!

File: mule-deployer/src/main/java/org/mule/tools/validation/DeploymentEnvironmentVersion.java (hunk 9 above)
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/validation/DeploymentEnvironmentVersion.java]: Done!

File: mule-deployer/src/main/java/org/mule/tools/validation/DeploymentValidatorFactory.java (hunk 3 above)
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/validation/DeploymentValidatorFactory.java]: Done!

File: mule-deployer/src/main/java/org/mule/tools/validation/DeploymentEnvironmentVersion.java (hunk 2 above)
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/validation/DeploymentEnvironmentVersion.java]: As we saw that this creates a cycle in the dependency graph I create a task to extract this functionality to a different module

File: mule-deployer/src/main/java/org/mule/tools/validation/agent/AgentDeploymentValidator.java (hunk 4 above)
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/validation/agent/AgentDeploymentValidator.java]: Done!

File: mule-deployer/src/main/java/org/mule/tools/validation/arm/ArmDeploymentValidator.java (hunk 5 above)
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/validation/arm/ArmDeploymentValidator.java]: Done!

File: mule-deployer/src/main/java/org/mule/tools/validation/cloudhub/CloudHubDeploymentValidator.java (hunk 6 above)
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/validation/cloudhub/CloudHubDeploymentValidator.java]: Done!

File: mule-deployer/src/main/java/org/mule/tools/validation/standalone/MuleCoreJarVersionFinder.java (hunk 7 above)
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/validation/standalone/MuleCoreJarVersionFinder.java]: Done!

**Pull request merged**